net: config: Add SYS_INIT handler to set clock from SNTP

pfalcon opened the pull request:
With this feature enabled (via CONFIG_NET_CONFIG_CLOCK_SNTP_INIT), a
POSIX application will automagically get correct absolute time via
calls to time(), gettimeofday(), etc.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: lib: config: Add SYS_INIT handler to set clock from SNTP

With this feature enabled (via CONFIG_NET_CONFIG_CLOCK_SNTP_INIT), an
application will automagically get correct absolute time via POSIX
functions like time(), gettimeofday(), etc.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
samples: posix: gettimeofday: Demo gettimeofday() returning correct time

Achieved using CONFIG_NET_CONFIG_CLOCK_SNTP_INIT option, i.e. querying
time over network using SNTP.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

jukkar commented on lib/posix/Kconfig:
```diff
@@ -96,6 +96,33 @@ config	POSIX_MAX_OPEN_FILES
+
+config POSIX_CLOCK_SNTP_INIT_SERVER
+	string "SNTP server to use for system clock init"
+	default "time.nist.gov"
```
The more I think about this, the more I get a feeling that we should not have some server name here by default. What about doing it like this:
```
config POSIX_CLOCK_SNTP_INIT_SERVER
	string "SNTP server to use for system clock init"
	default ""
	help
	    SNTP server name. See server information at
	    https://support.ntp.org/bin/view/Servers/NTPPoolServers to select a best NTP server
	    for your needs.
```

pfalcon replied:
But what are concerns with time.nist.gov exactly?

jukkar replied:
I remember few years back where some router vendor had hard coded some specific NTP server name into all of their routers. This caused lot of traffic to that specific server and bad publicity. Nothing against time.nist.gov, I think there is no issue regarding that specific server.
IMHO we should not promote any specific server here but let the end-user (be it individual hobbyist or a company) to do the necessary investigation and decision making, and force him/her to select the best server that would fit into their needs. Having a default value here means that most of the time nobody will do their due diligence about this matter.

pfalcon replied:
I see, ok, moving that setting to the sample...

pfalcon replied:
> The more I think about this, the more I get a feeling that we should not have some server name here by default. What about doing it like this:

Now done (with somewhat expanded help text).

pfalcon commented on lib/posix/init_clock_sntp.c:
```diff
@@ -0,0 +1,32 @@
+			      CONFIG_POSIX_CLOCK_SNTP_INIT_TIMEOUT, &ts);
+
+	if (res < 0) {
+		printk("ERROR: Cannot set current time using SNTP.\n");
```
> Why not have LOG_ERR() instead?

Done.

jukkar commented on lib/posix/init_clock_sntp.c:
```diff
@@ -0,0 +1,35 @@
+
+	struct sntp_time ts;
+	struct timespec tspec;
+	int res = sntp_simple(CONFIG_POSIX_CLOCK_SNTP_INIT_SERVER,
```
This will work only after networking is up, the IP address is set etc. So we might need here some more checks etc and possibly postpone this a bit after the net is up and running. To be investigated later when/if needed.

pfalcon replied:
> This will work only after networking is up, the IP address is set etc.

Yes, this is handled by carefully selecting default init priority to be after network auto-config, and mentioning this fact in Kconfig help.

One day, we might sit and review set of init priorities and see if we can make it "easier". (E.g. now, spacing is pretty tight between some levels.)

jukkar replied:
The issue will be seen if you have DHCP enabled, I am pretty sure this time setting will fail in that case.

jukkar reviewed (approved):
LGTM

pfalcon commented:
Now even README for sample is added ;-).

pfalcon commented:
Rebased.

dbkinder commented on samples/posix/gettimeofday/README.rst:
```diff
@@ -0,0 +1,32 @@
+Overview
+********
+
+This samples demonstrates POSIX gettimeofday() function working "out of box"
```
How about:
```
This  sample application demonstrates using the POSIX
gettimeofday() function to display the absolute wall clock
time every second. At system startup, the current time is
queried using the SNTP networking protocol, enabled by
setting the :option:`CONFIG_POSIX_CLOCK_SNTP_INIT`
and :option:`CONFIG_POSIX_CLOCK_SNTP_INIT_SERVER`
options.
```

pfalcon replied:
> How about:

Updated to use the suggested paragraph. @dbkinder, please re-review.

dbkinder reviewed (changes requested):
docs look good, with a suggested edit to the description.

nashif reviewed (changes requested):
This seems more like application code that ties posix subsystem with sntp from the networking subsystem. I do not think this code belongs to posix subsystem or more generally to zephyr. There can be different clock sources and this seems like a use case however one that should sit on top of zephyr, not within zephyr.

What makes sense however is to continue with the sample and provide everything as part of the sample.

dbkinder reviewed (approved):
+1 for doc changes

jukkar commented on lib/posix/Kconfig:
```diff
@@ -106,6 +110,39 @@ endif # FILE_SYSTEM
+	  wall clock time, and initialize system time from it, so
+	  functions like time(), gettimeofday(), etc. returned the
+	  correct absolute time (no just time since system startup).
+	  Requires networking.
```
BTW, as this requires networking, shouldn't we have `depends on NETWORKING && SNTP` here?

pfalcon replied:
> shouldn't we have depends on NETWORKING && SNTP here?

Ack, added.

pfalcon commented:
@nashif:

> This seems more like application code that ties posix subsystem with sntp from the networking subsystem. I do not think this code belongs to posix subsystem or more generally to zephyr.

This implements normative requirements from POSIX, specifically https://pubs.opengroup.org/onlinepubs/9699919799/functions/gettimeofday.html :

> The gettimeofday() function shall obtain the current time, expressed as seconds and microseconds since the Epoch, and store it in the timeval structure pointed to by tp.
---

@nashif :
> There can be different clock sources and this seems like a use case however one that should sit on top of zephyr, not within zephyr.

As a quick response, there can be different clock sources, and as a POSIX subsys maintainer, I welcome them all, for the purpose of adhering to POSIX normatives. In more detailed, there're 2 broad types of clock sources: board-builtin battery powered real-time clocks and networking time sources. This PR provides initial support for the 2nd class of source, using well-known (S)NTP protocol. This is about the only widely-used timing protocol in the industry. (But as soon as another protocol covering missing SNTP cases appears (e.g. some kind of authentication), we'll gladly support it.)

As for RTC, I'm not sure if we support those in Zephyr yet (I personally don't have any Zephyr-supported board with battery-powered RTC). But if/when the are, there certainly will be common driver model for them.

Summing up, if the concern that there will be too many clock source to support, that's not the case, as in foreseeable future, there're only 2 (SNTP and RTC drivers), with authenticated network time protocol in wishful thinking. We definitely have more advanced and even extravagant features "within" Zephyr. Again, here we talk about adhering to POSIX normative requirements. As for concerns re: "on top of", then Zephyr is itself a multi-level system, so we have design where some subsystem/lib is implemented on top of other, all "within" Zephyr.

> What makes sense however is to continue with the sample and provide everything as part of the sample.

Right, it's a shame that there're 0 samples for POSIX subsys. And this gettimeofday() sample sit in my tree for half year or so. I just remember getting a comment to one of previous POSIX PR that kinda providing empty compatibility headers for POSIX is "false advertizing". Or always remembered that looking at that gettimeofday() sample in my tree. Because the only thing it could demonstrate before is Zephyr's non-compliance with POSIX. Fortunately, with these changes, that's solved and we can demonstrate the bright side of Zephyr.

@nashif, So, please reconsider. Thanks.

pfalcon commented:
Added dev-review tag.

nashif reviewed (changes requested):
> Summing up, if the concern that there will be too many clock source to support, that's not the case, as in foreseeable future, the

This is not the main concern, the concern here is that we are mixing use-case and abstraction layers. The fact we have support for file systems does not mean we have to implement a file system as part of posix interface. SNTP or any other method to get the time should not be part of the posix layer, it is way up at the application layer. 
If you want this in zephyr, I have no issue with this being added as a sample showing how to glue SNTP and networking with the POSIX subsystem, but it should not be part of POSIX.

pfalcon commented:
> I have no issue with this being added as a sample showing how to glue SNTP and networking with the POSIX subsystem, but it should not be part of POSIX.

Well, it's not just a "sample", it's supposed to be useful, reusable part of the (POSIX) setup. It is put in lib/posix/ because it implements normative POSIX requirements, and to not breed more directories. If you think it would be better under e.g. lib/posix/clock_init/ , sounds good. If under lib/clock_init/, fine either, we already have lib/updatehub/, which I'd say more adhoc and peculiar featureset.

But hiding it somewhere under samples/ undermines my work throughout the year to improve POSIX compatibility and portability of existing POSIX software to Zephyr, and after so many efforts invested into it, I humbly don't agree with such mediocre finale. Again, proper timing source is a basic software requirement, and rather important one for networking software (no proper time, sensor reading are useless, but you can't even post them as TLS certificates fail to validate). Patching each and every software with adhoc Zephyr code, instead of just letting proper functionality be selected in config, - absolutely doesn't scale.

jukkar reviewed (changes requested):
@nashif has a good point here about SNTP stuff not belonging to POSIX layer. The code belongs more to application than to POSIX.

pfalcon commented:
>  The code belongs more to application than to POSIX.

To which application? It has nothing to do with application, it's system-level code.

I offered 2 alternatives above: a) move it to lib/posix/clock_init/; b) lib/clock_init/ . Let me know which is better (and better subdir names are welcome of course).

jukkar commented:
Why not integrate this with the `subsys/net/lib/config/init.c` which provides similar application level service, as a bonus that config init will handle the network startup and setting the time from there is then trivial.

pfalcon commented:
> Why not integrate this with the subsys/net/lib/config/init.c 

Hmm, as a separate module in subsys/net/lib/config/ , that doesn't seem a bad idea. @nashif, would you be ok with that?

nashif commented:
> Hmm, as a separate module in subsys/net/lib/config/ , that doesn't seem a bad idea. @nashif, would you be ok with that?

yes, works for me.

pfalcon commented:
Ok, thanks. Given the situation is clarified, removing dev-review label.

pfalcon committed:
include: posix: time.h: Make CLOCK_REALTIME, etc. compatible with Newlib

Otherwise, depending on the order of includes, different parts of code
may get different values for these constants.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Refactored based on the discussion above. @nashif, @jukkar, please re-review.

jukkar commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -0,0 +1,36 @@
+	return 0;
+}
+
+SYS_INIT(init_clock_via_sntp, APPLICATION, CONFIG_NET_CONFIG_SNTP_INIT_PRIO);
```
No need for this call. Just call `init_clock_via_sntp()` from init.c after the network is working (IP addresses set), so similar way as how network logging is initialized. Also place `init_clock_via_sntp()` directly to init.c. We also do not need separate priority as that is already handled by the init in config init.c.

pfalcon replied:
> Just call init_clock_via_sntp() from init.c after the network is working (IP addresses set), so similar way as how network logging is initialized.

Ok. I actually wanted to keep it more modular, re: discussions of possible extending that functionality to init via on-board RTC, etc., but ok, let's keep to the current functionality we have.

> Also place init_clock_via_sntp() directly to init.c.

Well, let's please keep having modular code, and not stuff everything into single gigantic file which then will be hard to work with. It's the same like existing bt_settings.c, ieee802154_settings.c - those could be put into init.c (they aren't too big either), but having them separate makes code structure much clearer.

jukkar reviewed (approved):
LGTM

pfalcon commented:
@jukkar: Thanks for approval. @nashif: Changed based on the discussion above, please re-review.
